### PR TITLE
chore(condo): DOMA-9105 QR code mutations return fields & logic

### PR DESCRIPTION
**Pull request by AleX83Xpert**
- [x] add acquiring host url to `validateQRCode` and `createPaymentByLink`
- [x] add currency code to both mutations
- [x] the logic of detection if receipt already paid should work similary for both mutations

### CODE REVIEW

File: apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js [hunk 1]
```diff
@@ -59,7 +56,7 @@ const CreatePaymentByLinkService = new GQLCustomSchema('CreatePaymentByLinkServi
         },
         {
             access: true,
-            type: 'type CreatePaymentByLinkOutput { multiPaymentId: ID!, amount: String!, explicitFee: String!, totalAmount: String!, integrationHostUrl: String!, address: String!, addressMeta: AddressMetaField!, unitType: String!, unitName: String!, accountNumber: String!, period: String! }',
+            type: 'type CreatePaymentByLinkOutput { multiPaymentId: ID!, amount: String!, explicitFee: String!, totalAmount: String!, acquiringIntegrationHostUrl: String!, currencyCode: String!, address: String!, addressMeta: AddressMetaField!, unitType: String!, unitName: String!, accountNumber: String!, period: String! }',
```
Comment (Alllex202) [inline apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:59]: Old mobile apps are not updated. Shouldn't we support the old API for this and not make any breaking changes?

File: apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js [hunk 2]
```diff
@@ -68,72 +65,31 @@ const CreatePaymentByLinkService = new GQLCustomSchema('CreatePaymentByLinkServi
...
-                for (const billingContext of billingContexts) {
-                    const orgId = get(billingContext, ['organization', 'id'])
-                    for (const acquiringContext of acquiringContexts) {
-                        if (orgId === get(acquiringContext, ['organization', 'id'])) {
-                            contextsMap[billingContext.id] = acquiringContext.id
-                        }
-                    }
-                }
+                const [organizationId] = Object.keys(auxiliaryData.contexts)
+                const { billingContext, acquiringContext } = auxiliaryData.contexts[organizationId]
```
Comment (Alllex202) [inline apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:87]: Why were all contexts used before, but now only the first one?

Comment (Alllex202) [reply apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:87]: Will there always be 1 context here?

File: apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js (hunk 1 above)
Comment (Alllex202) [inline apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:59]: Old mobile apps are not updated for us. Should we support the old API for this and not make any breaking changes?

File: apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js (hunk 2 above)
Comment (AleX83Xpert) [reply apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:87]: Here will be only one organization. If there are 2+ organization an error will be thrown by `validateQRCode` called above.

File: apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js (hunk 1 above)
Comment (AleX83Xpert) [reply apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:59]: For now both mutations are not used. Mobile applications asked for these changes.

File: apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js (hunk 2 above)
Comment (Alllex202) [inline apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:87]: throw error if no organization or context?

File: apps/condo/domains/billing/schema/ValidateQRCodeService.js [hunk 3]
```diff
@@ -131,44 +141,50 @@ const ValidateQRCodeService = new GQLCustomSchema('ValidateQRCodeService', {
...
                 const qrCodeAmount = String(Big(qrCodeFields.Sum).div(100))
-                const { PayeeINN } = qrCodeFields
+                const { PersAcc, PersonalAcc, PaymPeriod } = qrCodeFields
+                const period = formatPeriodFromQRCode(PaymPeriod)
 
-                const organizations = await Organization.getAll(context, {
-                    tin: PayeeINN,
-                    deletedAt: null,
-                })
+                const auxiliaryData = await findAuxiliaryData(qrCodeFields, { address: ERRORS.ADDRESS_IS_INVALID })
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/billing/schema/ValidateQRCodeService.js:147]: ¯\\\_(ツ)_/¯
Data returned from this function need for both mutation. Just following DRY.

File: apps/condo/domains/billing/schema/ValidateQRCodeService.js [hunk 4]
```diff
@@ -45,13 +43,13 @@ const ERRORS = {
         mutation: 'validateQRCode',
         code: INTERNAL_ERROR,
         type: NOT_FOUND,
-        message: 'Organization with provided TIN is not registered with Doma.ai',
+        message: 'Organization with provided TIN and having provided address is not registered with Doma.ai',
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/billing/schema/ValidateQRCodeService.js:46]: We do not need translations
image

File: apps/condo/domains/billing/schema/ValidateQRCodeService.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/billing/schema/ValidateQRCodeService.js:147]: Also, there is a js-doc

File: apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js (hunk 2 above)
Comment (AleX83Xpert) [reply apps/condo/domains/acquiring/schema/CreatePaymentByLinkService.js:87]: thrown in validateQRCode

File: apps/condo/domains/billing/utils/receiptQRCodeUtils.spec.js [hunk 5]
```diff
@@ -33,4 +72,174 @@ describe('receiptQRCodeUtils', () => {
...
+                expect(resolvers.onReceiptPeriodNewerThanQrCodePeriod).toBeCalledTimes(0)
+                expect(resolvers.onReceiptPeriodOlderThanQrCodePeriod).toBeCalledTimes(1)
+
+                expect(resolvers.onReceiptPeriodOlderThanQrCodePeriod).toHaveBeenCalledWith(billingReceiptForComparison)
+            })
+        })
+    })
+
+    test('Auxiliary data has correct format', async () => {
+        const [o10n] = await createTestOrganization(adminClient)
```
Comment (Alllex202) [inline apps/condo/domains/billing/utils/receiptQRCodeUtils.spec.js:200]: How about test a for the address error case?

File: apps/condo/domains/billing/schema/ValidateQRCodeService.js [hunk 6]
```diff
@@ -131,44 +141,50 @@ const ValidateQRCodeService = new GQLCustomSchema('ValidateQRCodeService', {
...
+
+                const billingContext = get(auxiliaryData, ['contexts', organizationId, 'billingContext'])
+                const acquiringContext = get(auxiliaryData, ['contexts', organizationId, 'acquiringContext'])
 
-                const acquiringContexts = await find('AcquiringIntegrationContext', {
-                    organization: { id_in: organizations.map((org) => org.id), deletedAt: null },
-                    status: CONTEXT_FINISHED_STATUS,
-                    deletedAt: null,
-                })
+                if (!acquiringContext) throw new GQLError(ERRORS.NO_ACQUIRING_CONTEXT, context)
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/billing/schema/ValidateQRCodeService.js]: The error and the test was added

File: apps/condo/domains/billing/schema/ValidateQRCodeService.test.js [hunk 7]
```diff
@@ -203,73 +204,70 @@ describe('ValidateQRCodeService', () => {
...
         test('should throw if found AcquiringIntegrationContext is not active (status !== Finished)', async () => {
-            const [integration] = await createTestAcquiringIntegration(adminClient)
-            const [acquiringContext] = await createTestAcquiringIntegrationContext(adminClient, organization, integration, { status: 'InProgress' })
+            const [anotherOrganization] = await createTestOrganization(adminClient)
+            const [anotherProperty] = await createTestProperty(adminClient, anotherOrganization)
+            await addBillingIntegrationAndContext(adminClient, anotherOrganization, {}, { status: CONTEXT_FINISHED_STATUS })
+            const anotherQrCodeObj = generateQrCodeObj({
+                PayeeINN: anotherOrganization.tin,
+                PayerAddress: `${anotherProperty.address}, кв. 1`,
+            })
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/billing/schema/ValidateQRCodeService.test.js]: Fixed

File: apps/condo/domains/billing/utils/receiptQRCodeUtils.spec.js [hunk 8]
```diff
@@ -1,8 +1,47 @@
-const { catchErrorFrom } = require('@open-condo/keystone/test.utils')
+/**
+ * @jest-environment node
+ */
+const index = require('@app/condo/index')
+const { faker } = require('@faker-js/faker')
+const Big = require('big.js')
 
-const { getQRCodeMissedFields, parseReceiptQRCode, formatPeriodFromQRCode } = require('./receiptQRCodeUtils')
+const { getById } = require('@open-condo/keystone/schema')
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/billing/utils/receiptQRCodeUtils.spec.js:8]: Yes, why not. I'd like data returned by `getById()` will be the function argument.

File: apps/condo/domains/billing/utils/receiptQRCodeUtils.spec.js (hunk 5 above)
Comment (AleX83Xpert) [reply apps/condo/domains/billing/utils/receiptQRCodeUtils.spec.js:200]: Added

File: apps/condo/domains/billing/schema/ValidateQRCodeService.js [hunk 9]
```diff
@@ -67,6 +71,18 @@ const ERRORS = {
...
+        code: BAD_USER_INPUT,
+        type: WRONG_FORMAT,
+        message: 'The provided address is invalid',
+    },
+    SEVERAL_ORGANIZATIONS: (count) => ({
+        mutation: 'validateQRCode',
+        code: INTERNAL_ERROR,
+        type: NOT_FOUND,
+        message: `Found ${count} with provided TIN and having provided address`,
+    }),
```
Comment (Alllex202) [inline apps/condo/domains/billing/schema/ValidateQRCodeService.js:85]: Could you please add a test for this error? :))

**Review by Alllex202 — APPROVED** [review]

**Pull request merged**